### ISSUE TITLE
Unable to load fips provider in release build using config from BIO buffer

### ISSUE THREAD

**Issue by skynet1381** [openssl/openssl#18484]
Hi,

For fips mode, we are trying to load fips provider from config using BIO buffer instead of from file.

Out fips config looks like below:

```
constexpr const char FipsConfig[] = R"(config_diagnostics = 1
openssl_conf = openssl_init

[openssl_init]
providers = provider_sect

[provider_sect]
base = base_sect
fips = fips_sect

[base_sect]
activate = 1

[fips_sect]
activate = 1
conditional-errors = 1
security-checks = 1
module-mac = CA:B4:A5:B1:09:35:FE:B2:2C:1E:0C:68:50:FF:55:F8:7D:F9:26:BB:6A:7B:3A:24:F5:77:4E:CD:A4:D9:4B:04)";
```

To load this config to our custom OSSL_LIB_CTX using BIO buffer, the pseudo flow of the code is:

```
 std::unique_ptr<CONF, CnfDeleter> cnf{
      NCONF_new_ex(SSL_lib_ctx.get(), nullptr)};
  if (!cnf) {
     log error;
  }

   std::unique_ptr<BIO, BioDeleter> bio{BIO_new_mem_buf(
        fips_config.c_str(), static_cast<int>(fips_config.length()))};
    if (!bio) {
      log error;
    }

   long errline{-1};
   auto ret = NCONF_load_bio(cnf.get(), bio.get(), &errline);
   if (ret <= 0) {
     log error;
   }

   ret = CONF_modules_load(cnf.get(), nullptr, 0);
   if (ret <= 0) {
       log error;
   }

```

At the end of this setup, I am calling below API to print and check which providers are loaded in my context:

```
OSSL_PROVIDER_do_all(SSL_lib_ctx.get(), testcallback, nullptr);
```

The issue I am seeing is in Debug build, I get both base and fips provider correctly loaded but in release mode, without encountering any error, there is only base provider gets loaded.

Also, just to clarify, we are always loading default provider from code so default provider is successfully there in both debug and release build. 

Can someone guide me to investigate this further?

**Comment by t8m** [comment]
The `(` and `)` in the config look suspicious. Also is the module mac the real SHA2-256 HMAC computed over the fips.so file with the right key?

**Comment by skynet1381** [comment]
The ( and ) are the start and end indicator of C++ literal strings. They are not part of config buffer.

I believe HMAC is correct as I have double checked using correct key. 

The same config in debug mode (with different module-mac for debug build) works fine. If there was any error in config or wrong module-mac, I was expecting some of the load API to fail and indicate error but that is not happening. My config load code above runs fine without any return error and still fips module is not loaded.

**Comment by mattcaswell** [comment]
Do you get anything useful in the OpenSSL error stack, e.g. via `ERR_print_errors_fp(stderr);`?

**Comment by skynet1381** [comment]
I do not see in error in stack during above config load operation. 

After the config load, later we do:

```
SSL_CTX *ctx = SSL_CTX_new_ex(SSL_lib_ctx.get(), "", TLS_method());
```

which gives below error stack:

```
crypto/evp/evp_fetch.c,inner_evp_generic_fetch,349: error:0308010C:digital envelope routines::unsupported
Non-default library context, Algorithm (EC : 88), Properties ()
crypto/evp/evp_fetch.c,inner_evp_generic_fetch,349: error:0308010C:digital envelope routines::unsupported
Non-default library context, Algorithm (EC : 88), Properties ()
crypto/evp/evp_fetch.c,inner_evp_generic_fetch,349: error:0308010C:digital envelope routines::unsupported
Non-default library context, Algorithm (EC : 88), Properties ()
crypto/evp/evp_fetch.c,inner_evp_generic_fetch,349: error:0308010C:digital envelope routines::unsupported
Non-default library context, Algorithm (EC : 88), Properties ()
 crypto/evp/evp_fetch.c,inner_evp_generic_fetch,349: error:0308010C:digital envelope routines::unsupported
Non-default library context, Algorithm (EC : 88), Properties ()
crypto/evp/evp_fetch.c,inner_evp_generic_fetch,349: error:0308010C:digital envelope routines::unsupported
Non-default library context, Algorithm (EC : 88), Properties ()
crypto/evp/evp_fetch.c,inner_evp_generic_fetch,349: error:0308010C:digital envelope routines::unsupported
Non-default library context, Algorithm (EC : 88), Properties ()
 crypto/evp/evp_fetch.c,inner_evp_generic_fetch,349: error:0308010C:digital envelope routines::unsupported
Non-default library context, Algorithm (X25519 : 91), Properties ()
 crypto/evp/evp_fetch.c,inner_evp_generic_fetch,349: error:0308010C:digital envelope routines::unsupported
Non-default library context, Algorithm (X448 : 92), Properties ()
crypto/evp/evp_fetch.c,inner_evp_generic_fetch,349: error:0308010C:digital envelope routines::unsupported
Non-default library context, Algorithm (DH : 86), Properties ()
crypto/evp/evp_fetch.c,inner_evp_generic_fetch,349: error:0308010C:digital envelope routines::unsupported
Non-default library context, Algorithm (DH : 86), Properties ()
crypto/evp/evp_fetch.c,inner_evp_generic_fetch,349: error:0308010C:digital envelope routines::unsupported
Non-default library context, Algorithm (DH : 86), Properties ()
crypto/evp/evp_fetch.c,inner_evp_generic_fetch,349: error:0308010C:digital envelope routines::unsupported
Non-default library context, Algorithm (DH : 86), Properties ()
crypto/evp/evp_fetch.c,inner_evp_generic_fetch,349: error:0308010C:digital envelope routines::unsupported
Non-default library context, Algorithm (DH : 86), Properties ()
ssl/ssl_lib.c,SSL_CTX_new_ex,3253: error:0A0000A1:SSL routines::library has no ciphers
```

**Comment by skynet1381** [comment]
Also, this is the call I am using to test which providers are loaded after above config load.

```
int testcallback(OSSL_PROVIDER *prov, void *) {
  print("Loaded provider name: ", OSSL_PROVIDER_get0_name(prov));
  return 1;
}

OSSL_PROVIDER_do_all(SSL_lib_ctx.get(), testcallback, nullptr);
```

and It logs below:

```
m=Loaded provider name: : v=base
m=Loaded provider name: : v=default
```

As I had mentioned below, default provider is loaded first from the code.

in the debug build, I see below log with same config structure:

```
m=Loaded provider name: : v=base
m=Loaded provider name: : v=default
m=Loaded provider name: : v=fips
```

**Comment by mattcaswell** [comment]
Could there be some environment difference between your debug and release builds?

You might want to try running your application through strace and look for where it loads the config file to see if that sheds any light.

**Comment by skynet1381** [comment]
I am trying to make sense of this but this is what I see with strace during the config load time. 
I see the fips.so is getting loaded in memory but then gets unloaded.

```
open("/home/devuser/build/mx_rel64/install/bin/../lib/fips.so", O_RDONLY) = 20
read(20, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0\0\220\1\0\0\0\0\0"..., 832) = 832
lseek(20, 1849264, SEEK_SET)            = 1849264
read(20, "\4\0\0\0\20\0\0\0\5\0\0\0GNU\0\2\0\0\300\4\0\0\0\3\0\0\0\0\0\0\0"..., 768) = 768
fstat(20, {st_mode=S_IFREG|0755, st_size=1947080, ...}) = 0
mmap(NULL, 4044072, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_DENYWRITE, 20, 0) = 0x7f3ec5a1e000
mprotect(0x7f3ec5be2000, 2097152, PROT_NONE) = 0
mmap(0x7f3ec5de2000, 94208, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 20, 0x1c4000) = 0x7f3ec5de2000
mmap(0x7f3ec5df9000, 1320, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x7f3ec5df9000
close(20)                               = 0
mprotect(0x7f3ec5de2000, 90112, PROT_READ) = 0
futex(0x7f3ec5df93f0, FUTEX_WAKE_PRIVATE, 2147483647) = 0
open("/home/devuser/build/mx_rel64/install/bin/../lib/fips.so", O_RDONLY) = 20
fstat(20, {st_mode=S_IFREG|0755, st_size=1947080, ...}) = 0
mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f3eca5cc000
read(20, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0\0\220\1\0\0\0\0\0"..., 4096) = 4096
read(20, " \335\27\0\0\0\0\0\0K<\0\0\0\0\0\10\0\0\0\0\0\0\0\326\330\27\0\0\0\0\0"..., 4096) = 4096
read(20, "\10\0\0\0\0\0\0\0\v\332\27\0\0\0\0\0\10U<\0\0\0\0\0\10\0\0\0\0\0\0\0"..., 4096) = 4096
read(20, "0d<\0\0\0\0\0\10\0\0\0\0\0\0\0\340\313\2\0\0\0\0\0@d<\0\0\0\0\0"..., 4096) = 4096
.
.
.
.
.
read(20, "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0"..., 4096) = 4096
read(20, "\270\235=\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\226\207\1\0\0\0\0\0"..., 4096) = 4096
read(20, "\0\0\0\0\0\0\0\0T\0\0\0\4\0\0\0\2\0\0\0\0\0\0\0\30\t\0\0\0\0\0\0"..., 4096) = 1480
read(20, "", 4096)                      = 0
read(20, "", 4096)                      = 0
close(20)                               = 0
munmap(0x7f3eca5cc000, 4096)            = 0
```

**Comment by paulidale** [comment]
I can confirm that this code works both in debug and production builds:
```c
#include <string.h>
#include <openssl/conf.h>
#include <stdlib.h>
#include <stdio.h>
#include <openssl/provider.h>
#include <openssl/err.h>
#include <openssl/bio.h>
#include <openssl/buffer.h>

static char *conf = "config_diagnostics = 1\n"
                   "openssl_conf = openssl_init\n"
                   "\n"
                   "[openssl_init]\n"
                   "providers = provider_sect\n"
                   "\n"
                   "[provider_sect]\n"
                   "base = base_sect\n"
                   "fips = fips_sect\n"
                   "\n"
                   "[base_sect]\n"
                   "activate = 1\n"
                   "\n"
                   "[fips_sect]\n"
                   "activate = 1\n"
                   "conditional-errors = 1\n"
                   "security-checks = 1\n"
                   "module-mac = E3:49:8D:DC:F9:55:46:DF:ED:69:5B:09:10:8B:EE:C9:91:EC:75:BD:0C:37:C3:71:85:EB:1C:FF:20:B0:1F:9C\n";

static int cb(OSSL_PROVIDER *p, void *data)
{
    printf("%s\n", OSSL_PROVIDER_get0_name(p));
    return 1;
}

int main() {
    long err = 0;
    CONF *c = NCONF_new(NULL);
    BIO *b = BIO_new_mem_buf(conf, strlen(conf));
    int ret = EXIT_FAILURE;
    
    if (b == NULL) goto err;
    if (conf == NULL) goto err;

    if (NCONF_load_bio(c, b, &err) <= 0) goto err;
    if (CONF_modules_load(c, NULL, 0) <= 0) goto err;
    OSSL_PROVIDER_do_all(NULL, cb, NULL);
    ret = EXIT_SUCCESS;
err:
    if (ret != EXIT_SUCCESS)
        ERR_print_errors_fp(stdout);
    NCONF_free(c);
    BIO_free(b);
    return EXIT_FAILURE;
}
```
I'd suggest also add the legacy provider into the mix since it is _easier_ to load than the FIPS one -- it not loading points you to an environmental problem rather than a checksum one.

I suspect your not getting the provider paths set up properly and the FIPS provider isn't being found.

**Comment by mattcaswell** [comment]
> I suspect your not getting the provider paths set up properly and the FIPS provider isn't being found.

strace seems to suggest that it is found but subsequently fails. This suggests a checksum problem.

Attempting to load the legacy provider instead of the fips provider is a good suggestion. This will tell us if there is a fundamental problem loading providers at all, or if it is specific to the fips module.

**Comment by skynet1381** [comment]
legacy provider is loading fine. So as you suggested I am checking with our symbol stripping code to make sure that checksum is not changed.

**Comment by skynet1381** [comment]
yes, that was it. I thought I had correct checksum but after that we had a step to strip symbols from fips.so which would change the checksum.

Thanks for helping with this. 

I think a proper return error or on error stack for the wrong checksum would be very helpful. I think I have seen checksum error when using config load from file but this path of loading from buffer is not giving the error.

**Comment by nhorman** [comment]
marking as inactive, to be closed at the end of 3.4 dev barring further input
